Mathlib/FieldTheory/IsPerfectClosure.lean

```diff
@@ -0,0 +1,421 @@
+    pNilradical R p = ⊥ := pNilradical_eq_bot (not_lt.2 hp)
+
+theorem mem_pNilradical {R : Type*} [CommSemiring R] {p : ℕ} {x : R} :
+    x ∈ pNilradical R p ↔ ∃ n : ℕ, x ^ p ^ n = 0 := by
+  by_cases hp : 1 < p
+  · rw [pNilradical_eq_nilradical hp]
+    refine ⟨fun ⟨n, h⟩ ↦ ⟨n, ?_⟩, fun ⟨n, h⟩ ↦ ⟨p ^ n, h⟩⟩
+    rw [← Nat.sub_add_cancel ((Nat.lt_pow_self hp n).le), pow_add, h, mul_zero]
+  rw [pNilradical_eq_bot hp, Ideal.mem_bot]
+  refine ⟨fun h ↦ ⟨0, by rw [pow_zero, pow_one, h]⟩, fun ⟨n, h⟩ ↦ ?_⟩
+  rcases Nat.le_one_iff_eq_zero_or_eq_one.1 (not_lt.1 hp) with hp | hp
+  · by_cases hn : n = 0
+    · rwa [hn, pow_zero, pow_one] at h
+    rw [hp, zero_pow hn, pow_zero] at h
+    haveI := subsingleton_of_zero_eq_one h.symm
+    exact Subsingleton.elim _ _
+  rwa [hp, one_pow, pow_one] at h
+
+section IsPerfectClosure
+
+variable {K L M : Type*}
+
+section CommSemiring
+
+variable [CommSemiring K] [CommSemiring L] [CommSemiring M]
+  (i : K →+* L) (j : K →+* M) (p : ℕ) [ExpChar K p] [ExpChar L p] [ExpChar M p]
+
+/-- If `i : K →+* L` is a ring homomorphism of characteristic `p` rings, then it makes `L`
+a perfect closure of `K` if the following conditions are satisfied:
+
+- `L` is a perfect ring.
+- `i : K →+* L` is "purely inseparable" in the sense that for any element `x` of `L` there exists
+  a natural number `n` such that `x ^ (p ^ n)` is contained in `K`.
+- The kernel of `i` is contained in (in fact, equal to, see `IsPerfectClosure.ker_eq`)
+  the `p`-nilradical of `K`. -/
+@[mk_iff]
+class IsPerfectClosure : Prop where
+  perfectRing' : PerfectRing L p
+  pow_mem' : ∀ x : L, ∃ (n : ℕ) (y : K), i y = x ^ p ^ n
+  ker_le' : RingHom.ker i ≤ pNilradical K p
+
+theorem IsPerfectClosure.perfectRing [IsPerfectClosure i p] :
+    PerfectRing L p := perfectRing' i
+
+theorem IsPerfectClosure.pow_mem [IsPerfectClosure i p] (x : L) :
+    ∃ (n : ℕ) (y : K), i y = x ^ p ^ n := pow_mem' x
+
+theorem IsPerfectClosure.ker_le [IsPerfectClosure i p] :
+    RingHom.ker i ≤ pNilradical K p := ker_le'
+
+/-- If `i : K →+* L` is a ring homomorphism of exponential characteristic `p` rings, such that `L`
+is perfect, then the `p`-nilradical of `K` is contained in the kernel of `i`. -/
+theorem RingHom.pNilradical_le_ker_of_perfectRing [PerfectRing L p] :
+    pNilradical K p ≤ RingHom.ker i := fun x h ↦ by
+  obtain ⟨n, h⟩ := mem_pNilradical.1 h
+  replace h := congr((frobeniusEquiv L p).symm^[n] (i $h))
+  rwa [map_pow, ← iterate_frobenius, iterate_frobeniusEquiv_symm_apply_iterate_frobenius,
+    map_zero, iterate_frobeniusEquiv_symm_zero] at h
+
+theorem IsPerfectClosure.ker_eq [IsPerfectClosure i p] : RingHom.ker i = pNilradical K p :=
+  ker_le'.antisymm (haveI := perfectRing i p; i.pNilradical_le_ker_of_perfectRing p)
+
+/-- A perfect ring is its perfect closure. -/
+instance IsPerfectClosure.self_of_perfect [PerfectRing M p] :
+    IsPerfectClosure (RingHom.id M) p where
+  perfectRing' := ‹_›
```

Comment:
I assume that `[perfectRing' : PerfectRing L p]` will give you an `instance IsPerfectClosure.perfectRing'`, right? But this will not quite work if I understand correctly.

Reply:
Does it really give the instance, and what do you mean by "not quite work"? I think `extends` will give the instance, [] should just auto-fill the instance by typeclass inference.

Reply:
> Does it really give the instance, and what do you mean by "not quite work"?

I mean `instance IsPerfectClosure.perfectRing` will not work because LHS has more free variables than RHS. Only `theorem IsPerfectClosure.perfectRing` works.

> [] should just auto-fill the instance by typeclass inference.

That's good to know, which is exactly what I wanted.